scipy/optimize/cython_optimize/__init__.py

```diff
@@ -0,0 +1,630 @@
+"""
+Cython Optimize Zeros API
+=========================
+This package contains a Cython API for the root finding scalar functions in
+:py:mod:`scipy.optimize` to obtain efficient pure C implementations, although
+faster execution is not guaranteed. For example the Cython Optimize Zeros API
+could be used for native C looping or with Cython ``prange`` for efficient
+execution of large arrays of inputs with the same callbacks. The Cython
+Optimize Zeros API can be imported into Cython code using the following::
+
+    cimport scipy.optimize.cython_optimize as coz
+
+Callback Signatures
+-------------------
+The Cython Optimize Zeros API provides three callback signagures for the root
+finders, two which are safe without the global interpreter lock (GIL) and one
+that uses Python tuples. The functions are grouped by callback signatures into
+Cython modules that can be imported using ``cimport``.
+
+.. py:module:: scipy.optimize.cython_optimize.zeros_struct
+
+``cython_optimize.zeros_struct``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The root-finders in this module use a callback that takes a double with the
+scalar independent variable as the 1st argument and a user defined ``struct``
+with any extra parameters as the 2nd. ::
+
+    double (*callback_type)(double, void*)
+
+This module can be imported using the following::
+
+    from scipy.optimize.cython_optimize cimport zeros_struct
+
+.. py:module:: scipy.optimize.cython_optimize.zeros_array
+
+``cython_optimize.zeros_array``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The root-finders in this module use a callback that takes an integer with the
+number of extra parameters as the 1st argument and an array of doubles with any
+extra parameters as the 2nd. Even if the integer is unused in your callback, it
+must still be in the signature, because an internal wrapper will use it to
+prepend the scalar independent variable to the array, which is then passed to
+your callback. Therefore, in your callback, the independent variable must be
+the first element in the array, followed by the extra parameters.  Also the
+maximum number of extra parameters is hard-coded as ``MAXARGS = 10``. ::
+
+    double (*callback_type_array)(int, double*)
+
+This module can be imported using the following::
+
+    from scipy.optimize.cython_optimize cimport zeros_array
+
+.. py:module:: scipy.optimize.cython_optimize.zeros_tuple
+
+``cython_optimize.zeros_tuple``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The root-finders in this module use a callback that takes a double with the
+scalar independent variable as the 1st argument and a Python tuple with any
+extra parameters as the 2nd. Therefore this signature is *not* safe to use with
+``nogil``. ::
+
+    double (*callback_type_tuple)(double, tuple)
+
+This module can be imported using the following::
+
+    from scipy.optimize.cython_optimize cimport zeros_tuple
+
+Available Functions
+-------------------
+There are seven available functions which are all implemented in pure C and
```

Comment:
It's confusing to say 7 and then have a list of 5 below (I know why, `newton` has 3 methods). Probably better to match the Python and Cython APIs here. (EDIT: that's explained below - not clear it's better to make these different)